### ISSUE TITLE
[1.0.3] Cython not called for missing .c files

### ISSUE THREAD

**Issue by olebole** [astropy/astroscrappy#6]
Hi, I am [currently preparing](https://anonscm.debian.org/cgit/debian-astro/packages/astroscrappy.git) a [Debian package](http://bugs.debian.org/804520) for astroscrappy; and this needs to rebuild everything from scratch (invoking cython for the `*.pyx` files). To ensure this, I removed the three .c files before the build. Cython and cython3 are installed.

The build process however fails with the error

```
running build_ext
error: [Errno 2] Could not find C/C++ file astroscrappy/utils/median_utils.(c/cpp)
    for Cython file astroscrappy/utils/median_utils.pyx
    when building extension astroscrappy.utils.median_utils.
    Cython must be installed to build from a git checkout.: 'astroscrappy/utils/median_utils.c'
E: pybuild pybuild:274: build: plugin distutils failed with:
    exit code=1: /usr/bin/python setup.py build 
```

Otherwise, the package builds fine.

Best regards

Ole

**Comment by cmccully** [comment]
@olebole That's great!

There are a few C files that are not Cython files: imutils.h, imutils.c, medutils.c, medutils.h are all not generated by Cython, so don't delete those.

What does the output look like for the "build" step of setup.py? Does Cython run?

What version of Python are you running?

**Comment by olebole** [comment]
Sure, I only delete those with `.pyx` files. Actually, the way to reproduce it is

 * install cython
 * delete `astroscrappy.c`, `utils/image_utils.c`, `utils/median_utils.c`
 * run `python setup.py build`

This gives the error message shown in the initial post. For this, I am using python-2.7.10.

**Comment by cmccully** [comment]
I am having trouble reproducing the error. I created a clean environment:
conda create -n "test_astroscrappy"  python=2.7 cython astropy numpy
Then I cloned this repo into a fresh directory and typed 
python setup.py install
which ran run build_ext as one of the steps.
Everything installed as I would expect on my Centos machine. 

Can you send the full log to see if something went wrong earlier? Do you have numpy installed beforehand?

**Comment by olebole** [comment]
I created a [gist with the output](https://gist.github.com/olebole/14ce5723ae477853209f) and the commands to reproduce them. This was however 1.0.3; maybe something in between fixed this already?
Numpy is installed.

**Comment by cmccully** [comment]
That's very strange. When I pull the 1.0.3 version from the git version, it builds correctly. However, when I pull from the pypi I can produce the error you are seeing. Is is possible to use a tar file created from only files contained in the github? 
I suspect this error is coming from deleting the files after the release scripts have been run. @eteq @embray  have you seen anything like this before?

**Comment by embray** [comment]
Try also deleting `astroscrappy/cython_version.py*`.  The relevant logic is here (though I'm open to suggestions to improving it): https://github.com/astropy/astropy-helpers/blob/master/astropy_helpers/commands/build_ext.py#L14

If `should_build_with_cython` returns False it won't pass the Cython sources through cythonize.

**Comment by cmccully** [comment]
Thanks @embray! That seems to have fixed it in my build. @olebole, does that fix things for you?

**Comment by olebole** [comment]
Yes it does, thanks. However, I don't understand the idea of this logic. If cython is there, the .c files could always just be rebuilt, right? (this discussion maybe, however, out of the scope of this specific issue)

**Comment by astrofrog** [comment]
@olebole - we ship the C files to avoid having to require Cython as a dependency.

**Comment by olebole** [comment]
I understand that -- it is not against shipping the C files.
However if it appears that cython is installed, you could just ignore the shipped C files and rebuild them. This would also help if someone wants to patch the .pyx files before the actual build.

**Comment by cmccully** [comment]
@olebole Are we satisfied with this answer? Should we close this issue?

**Comment by olebole** [comment]
That depends on you :-) First deleting the files works for me; in that sense it is fine. I would however consider this as a workaround and not as a  fix -- IMO it should also work without that (so, if there is cython installed, the shipped `.c` files could be just ignored and rebuild). But, as I said: I can live perfectly with the current solution.

**Comment by cmccully** [comment]
Is this something that should be fixed here or upstream in the package template? I think I am inclined to move this issue up the chain and then pull in the fixes back down.

**Comment by olebole** [comment]
No idea, it looks that some astropy affiliated packages (specutils?) have this problem -- maybe @embray could comment on this?

**Comment by cmccully** [comment]
We believe this should be resolved in #46.
